### PR TITLE
[FSDP] utility to apply optimizer during backward

### PR DESCRIPTION
**Pull request by robieta**
## What does this PR do?

Utility to apply optimizer in the backward pass. (Cleaned up version of what I showed in [my talk](https://www.youtube.com/watch?v=qRZrVNNe3gQ)) I've tried to be as principled as possible, but there's no getting away from the fact that FSDP is very complicated and thus this utility is also complicated. I'm also open to feedback on the test; I had to make it quite involved to cover all of the behavior, but if there are pre-existing utilities or better ways to do things feel free to point me to them.

Unlike the LLaMA demo where I could create an optimizer per Parameter, the general utility can't impose any such structural requirement. So instead the backward hook has to keep a record of which parameters it has seen and trigger the optimizer when all parameters are ready.

It will be better when this is in PyTorch core (I'm using TONS of non-public APIs), but as a caveat emptor proof of concept it works well enough.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/utilities/optimizer.py [hunk 1]
```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
...
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+                post_step()
+
+    try:
+        hook_handles = []
+        for h in param_handles:
+            assert isinstance(h, FlatParamHandle)
+            flat_param = h.flat_param
+            assert hasattr(flat_param, "_post_backward_hook_state")
```
Comment (justusschock) [inline src/lightning/fabric/utilities/optimizer.py:115]: is this just for mypy? otherwise an attribute error from the next line would be more specific than the assertion error here. If you need this for mypy, it is fine though :)

**Review by justusschock — APPROVED** [review]
looks very good to me. One very mjnor comment!

File: src/lightning/fabric/utilities/optimizer.py [hunk 2]
```diff
@@ -12,15 +12,36 @@
...
+        from torch.distributed.fsdp._common_utils import _get_module_fsdp_state
+        from torch.distributed.fsdp._traversal_utils import _get_fsdp_handles
+        from torch.distributed.fsdp.flat_param import FlatParameter, FlatParamHandle
+
+        SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD = True
+
+    # Note: Not `ImportError` because sometimes PyTorch has internal failures
+    # (such as AttributeError) if it was built without distributed support.
+    except BaseException:
+        pass
```
Comment (Borda) [inline src/lightning/fabric/utilities/optimizer.py:43]:
```suggestion

    # Note: Not `ImportError` because sometimes PyTorch has internal failures
    # (such as AttributeError) if it was built without distributed support.
    with contextlib.suppress(BaseException):
        from torch.distributed.fsdp._common_utils import _get_module_fsdp_state
        from torch.distributed.fsdp._traversal_utils import _get_fsdp_handles
        from torch.distributed.fsdp.flat_param import FlatParameter, FlatParamHandle

        SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD = True
```

File: src/lightning/fabric/utilities/optimizer.py [hunk 3]
```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
...
+    By moving optimizer step invocation into the backward call we can free
+    gradients earlier and reduce peak memory.
+    """
+    assert SUPPORTS_OPTIMIZER_IN_FSDP_BACKWARD
+    apply_lock = threading.Lock()
+
+    from lightning.fabric.wrappers import _FabricModule
+
+    assert isinstance(module_wrapper, _FabricModule)
+    param_handles = _get_fsdp_handles(module := module_wrapper._forward_module)
```
Comment (carmocca) [inline src/lightning/fabric/utilities/optimizer.py:81]: Could we have a protected pytorch-only implementation that doesn't rely on fabric and a public (but experimental) implementation that does? For example:


```py
def _apply_optimizers_during_fsdp_backward():
    ...

def fsdp_overlap_step_with_backward(fabric_module, optimizers):
    return _apply_optimizers_during_fsdp_backward(optimizers, fabric_module._forward_module)

```

File: src/lightning/fabric/utilities/optimizer.py [hunk 4]
```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
...
+                    remaining[p] = (optimizer, unfinished)
+
+    def maybe_step(parameters: Iterable[torch.nn.Parameter], post_step: Callable[[], None] = lambda: None) -> None:
+        for p in tuple(parameters):
+            optimizer, unfinished = remaining.pop(p)
+            unfinished.pop(p)
+            if not unfinished:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
+                post_step()
```
Comment (carmocca) [inline src/lightning/fabric/utilities/optimizer.py:108]: What is a good example of using this hook? Can you add assertions that it gets called as expected?

What if we remove it and the default step implementation is

```python
def step(optimizer):
    optimizer.step()
    optimizer.zero_grad()
```

and the user can do

```python
with _apply_optimizers_during_fsdp_backward(optimizers, model, my_step):
```

**Review by carmocca — COMMENTED** [review]
Very cool!

File: src/lightning/fabric/utilities/optimizer.py (hunk 1 above)
Comment (robieta) [reply src/lightning/fabric/utilities/optimizer.py]: Done.

File: src/lightning/fabric/utilities/optimizer.py [hunk 5]
```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
...
+                    unfinished[p] = None
+                    remaining[p] = (optimizer, unfinished)
+
+    def maybe_step(parameters: Iterable[torch.nn.Parameter], post_step: Callable[[], None] = lambda: None) -> None:
+        for p in tuple(parameters):
+            optimizer, unfinished = remaining.pop(p)
+            unfinished.pop(p)
+            if not unfinished:
+                optimizer.step()
+                optimizer.zero_grad(set_to_none=True)
```
[the remark this answers is not in the document]
Comment (robieta) [reply src/lightning/fabric/utilities/optimizer.py]: I was trying to be explicit, but ultimately I just went ahead and removed it.

File: src/lightning/fabric/utilities/optimizer.py [hunk 6]
```diff
@@ -31,4 +52,104 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
...
-        optimizer.state[p] = apply_to_collection(v, Tensor, move_data_to_device, device, allow_frozen=True)
+        optimizer.state[p] = apply_to_collection(v, torch.Tensor, move_data_to_device, device, allow_frozen=True)
+
+
+def _no_op() -> None:
+    pass
+
+
+@contextlib.contextmanager
+def _apply_optimizers_during_fsdp_backward(
```
[the remark this answers is not in the document]
Comment (robieta) [reply src/lightning/fabric/utilities/optimizer.py]: Done.

File: tests/tests_fabric/strategies/test_fsdp.py [hunk 7]
```diff
@@ -287,3 +295,161 @@ def test_set_timeout(init_process_group_mock):
...
+    def make_model_and_optimizers():
+        torch.manual_seed(0)
+
+        with fabric.device:
+            backbone = [Block(feature_dim) for _ in range(num_blocks)]
+            model = nn.Sequential(*backbone, nn.Linear(feature_dim, 1, bias=False))
+            optimizers = [torch.optim.SGD(layer.parameters(), lr=0.1, momentum=0.9) for layer in model]
+
+        model = fabric.setup_module(model)
+        optimizers = [fabric.setup_optimizers(optimizer) for optimizer in optimizers]
```
[the remark this answers is not in the document]
Comment (robieta) [reply tests/tests_fabric/strategies/test_fsdp.py]: I learned the hard way that it's `fabric.setup_optimizers(*optimizers)` ^^

File: tests/tests_fabric/strategies/test_fsdp.py [hunk 8]
```diff
@@ -287,3 +295,161 @@ def test_set_timeout(init_process_group_mock):
...
+    fabric = Fabric(accelerator="cuda", devices=num_gpus, strategy=strategy)
+    fabric.launch()
+
+    def make_model_and_optimizers():
+        torch.manual_seed(0)
+
+        with fabric.device:
+            backbone = [Block(feature_dim) for _ in range(num_blocks)]
+            model = nn.Sequential(*backbone, nn.Linear(feature_dim, 1, bias=False))
+            optimizers = [torch.optim.SGD(layer.parameters(), lr=0.1, momentum=0.9) for layer in model]
```
[the remark this answers is not in the document]
Comment (robieta) [reply tests/tests_fabric/strategies/test_fsdp.py]: Done.

File: src/lightning/fabric/utilities/optimizer.py (hunk 3 above)
Comment (robieta) [reply src/lightning/fabric/utilities/optimizer.py]: Great point!

File: src/lightning/fabric/utilities/optimizer.py (hunk 4 above)
Comment (robieta) [reply src/lightning/fabric/utilities/optimizer.py]: It's not user facing. It's an implementation detail to free memory when the step runs.

**Review by lantiga — APPROVED** [review]
Great work!
